rustls/src/msgs/handshake.rs

```diff
@@ -1524,6 +1533,39 @@ impl Codec<'_> for EcParameters {
     }
 }
 
+#[derive(Debug)]
+pub(crate) enum ClientKeyExchangeParams {
+    Ecdh(ClientEcdhParams),
+    Dh(ClientDhParams),
+}
+
+impl ClientKeyExchangeParams {
+    #[cfg(feature = "tls12")]
+    pub(crate) fn pub_key(&self) -> &[u8] {
+        match self {
+            Self::Ecdh(ecdh) => &ecdh.public.0,
+            Self::Dh(dh) => &dh.public.0,
+        }
+    }
+
+    #[cfg(feature = "tls12")]
+    pub(crate) fn encode(&self, buf: &mut Vec<u8>) {
+        match self {
+            Self::Ecdh(ecdh) => ecdh.encode(buf),
+            Self::Dh(dh) => dh.encode(buf),
+        }
+    }
+}
+
+impl KxDecode<'_> for ClientKeyExchangeParams {
+    fn decode(r: &mut Reader, algo: KeyExchangeAlgorithm) -> Result<Self, InvalidMessage> {
+        Ok(match algo {
+            KeyExchangeAlgorithm::ECDHE => Self::Ecdh(Codec::read(r)?),
```

Comment:
Style nit: instead of calling `Codec::read()`, I think this would be easier to follow if you wrote them as `ClientEcdhParams::read()`/`ClientDhParams::read()`.

(Maybe consider writing `use KeyExchangeAlgorithm::*;` before the `match` to cut down on noise.)